src/services/messageBoard.ts

```diff
@@ -0,0 +1,136 @@
+import { Pool } from "pg";
+import { Request, Response } from "express";
+import config from "config";
+
+const addressesRequestLimit: number = config.get("server.addressRequestLimit");
+export interface Message {
+  [key: string]: {
+    block_no: number;
+    title: string;
+    content: string;
+    valid: number | null;
+    expire: number | null;
+  };
+}
+
+export interface MessageJson {
+  [key: string]: {
+    title: string;
+    content: string[];
+    valid: number | null;
+    expire: number | null;
+  };
+}
+
+interface Dictionary<T> {
+  [keys: string]: T;
+}
+
+export const handleMessageBoard =
+  (p: Pool) =>
+  async (req: Request, res: Response): Promise<void> => {
+    if (!req.body.poolIds) throw new Error("No poolIds in body");
+    const hashes = req.body.poolIds;
+
+    if (!(hashes instanceof Array) || hashes.length > addressesRequestLimit)
+      throw new Error(
+        ` poolIds must have between 0 and ${addressesRequestLimit} items`
+      );
+
+    const fromBlock = req.body.fromBlock;
+    const untilBlock = req.body.untilBlock;
+
+    const ret: Dictionary<null | Message[]> = {};
+
+    for (const hash of hashes) {
+      if (hash.length !== 56) {
+        throw new Error(`Received invalid pool id: ${hash}`);
+      }
+
+      const queryMessageBoard = `
+      SELECT
+        b.block_no AS "block_no",
+        txm.json AS "message_json"
+      FROM tx_metadata txm
+        JOIN tx ON (txm.tx_id = tx.id)
+        JOIN block b ON (tx.block_id = b.id)
+        JOIN tx_out txo ON (txm.tx_id = txo.tx_id)
+      WHERE txo.stake_address_id IN (
+          SELECT id
+          FROM stake_address sa
+          WHERE (
+              RIGHT(
+                sa.hash_raw::VARCHAR,
+                LENGTH(sa.hash_raw::VARCHAR) -4
+              )
+            ) IN (
+              SELECT encode(po.hash, 'hex')
+              FROM pool_owner po
+                JOIN pool_hash ph ON (po.pool_hash_id = ph.id)
+              WHERE registered_tx_id = (
+                  SELECT MAX(registered_tx_id)
+                  FROM pool_owner po
+                    JOIN pool_hash ph ON (po.pool_hash_id = ph.id)
+                  WHERE encode(ph.hash_raw, 'hex') = $1
+                )
+            )
+        )
+        AND txm.key = 1990
+        AND (
+          $2::INTEGER IS NULL
+          OR b.block_no >= $2
+        )
+        AND (
+          $3::INTEGER IS NULL
+          OR b.block_no <= $3
+        )
+      ORDER BY tx.id DESC;
+    `;
```

Comment:
If I understand the query correctly, you are getting all metadata entries, joined by the corresponding txs, and then joined by the URxOs where the tx ID in the UTxO matches the tx ID of the metadata and then you filter all that by only those UTxOs where the staking key matches that of the pool owner.

If I understand everything correctly, I don't think this matches the CIP spec for the message board:

```
Message board communication is a type of metadata that has been included in an on-chain transaction between two base addresses associated with a stake pool operator owner address. Given the onetime fee for this communication, we are considering this as a message board of a stake pool, as it also enables delegates to easier access historical metadata communication.
```

The board message tx must be sent from tye pool owner to their own address (or at least with the same staking key) so on the query level we should be checking that **inputs and outputs have the same owner staking key AND all outputs have the same owner staking key**.

Because if a pool owner sends a tx to someone else's wallet , which could be a direct message , the change output would still have the owner's staking key and this query would identify it as a board message.

The query also only uses tx outputs right now, but it should be separately checking tx inputs and outputs. For inputs you need to join the `tx_in` table and then join the `tx_out` again through the `tx_in` as the `tx_in` only contains the indexes matching the outputs of previous transactions as inputs in next transactions.

Safe check for a match then, imo, would be to assert that ALL inputs and ALL outputs in a tx have the same matching staking key.

Reply:
This one's tricky.

> The board message tx must be sent from the pool owner to their own address (or at least with the same staking key) so on the query level we should be checking that inputs and outputs have the same owner staking key AND all outputs have the same owner staking key.

Yes you're right, I did add additional checks / removals of possibly misidentified messages.

> Because if a pool owner sends a tx to someone else's wallet , which could be a direct message , the change output would still have the owner's staking key and this query would identify it as a board message.

This should in fact not happen as direct message is identified by a different key (`1991`), but you're right about the change address. Should be covered now.

> The query also only uses tx outputs right now, but it should be separately checking tx inputs and outputs. For inputs you need to join the tx_in table and then join the tx_out again through the tx_in as the tx_in only contains the indexes matching the outputs of previous transactions as inputs in next transactions.

You;re right, but I don't know what you meant by separately, I've tried to do it in a single query so hopefully this one's covered as well.

Reply:
Feel free to reopen any of these if you feel it's not addressed properly. Moreover, I've rewritten the queries so that they would match the new (11) dbsync schema.

Reply:
> I don't know what you meant by separately, I've tried to do it in a single query so hopefully this one's covered as well.

Ah yes, I just meant that all inputs should be checked as well as all outputs, so just bad wording from my part )